pkg/common/health/cache.go

```diff
@@ -0,0 +1,186 @@
+package health
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/sirupsen/logrus"
+)
+
+type CheckState struct {
+	// Err is the error returned from a failed health check
+	Err error
+
+	// Details contains more contextual detail about a
+	// failing health check.
+	Details State
+
+	// CheckTime is the time of the last health check
+	CheckTime time.Time
+
+	// ContiguousFailures the number of failures that occurred in a row
+	ContiguousFailures int64
+
+	// TimeOfFirstFailure the time of the initial transitional failure for
+	// any given health check
+	TimeOfFirstFailure time.Time
+}
+
+func NewCache(log logrus.FieldLogger, clock clock.Clock) *Cache {
+	return &Cache{
+		allowedChecks: make(map[string]Checkable),
+		currentStatus: make(map[string]CheckState),
+		log:           log,
+		clk:           clock,
+	}
+}
+
+type Cache struct {
+	allowedChecks map[string]Checkable
+
+	currentStatus map[string]CheckState
+	mutex         sync.Mutex
+	clk           clock.Clock
+
+	log   logrus.FieldLogger
+	hooks struct {
+		statusUpdated chan struct{}
+	}
+}
+
+func (c *Cache) SetStatusUpdatedHook(statusUpdated chan struct{}) {
+	c.hooks.statusUpdated = statusUpdated
+}
+
+func (c *Cache) AddCheck(name string, checkable Checkable) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if _, ok := c.allowedChecks[name]; ok {
+		return fmt.Errorf("check %q has already been added", name)
+	}
+
+	c.allowedChecks[name] = checkable
+	return nil
+}
+
+func (c *Cache) GetStatuses() map[string]CheckState {
```

Comment:
It is dangerous to return a reference type that is protected by a lock since the caller can then mutate the reference object without going through the lock. Although unlikely this will be a problem if the map is mutated while the checker is reading the map.